### PR TITLE
HTTP2: Add DumpState support for HTTP2

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: HTTP2: Add DumpState support for HTTP2
Additional Description: While dispatching, the HTTP2 Connection object registers itself as a scopetrackedobject. If we crash during parsing pipeline, we'll dump information such as the current streams, current buffer, etc.
Risk Level: medium
Testing:unit tests
Docs Changes: N/A
Release Notes: Included
Platform Specific Features: N/A
Related Issue #14249

This is a follow up to PR #14812.

Originally got some feedback on this in https://github.com/KBaichoo/envoy/pull/2.

### CODE REVIEW

**Review by antoniovicente — COMMENTED** [review]
I think it looks good otherwise.

File: test/common/http/http2/codec_impl_test.cc
Comment (antoniovicente) [inline test/common/http/http2/codec_impl_test.cc:1100]: Make sure to adjust the \0's after you start escaping null chars.  I don't trust the code above to not treat the first \0 in this string literal as an end-of-c-string indicator

File: test/common/http/http2/codec_impl_test.cc
Comment (antoniovicente) [inline test/common/http/http2/codec_impl_test.cc:44]: This "using" is breaking the windows build.

2021-02-03T04:44:36.3165016Z   C:/Program Files (x86)/Microsoft Visual Studio/2019/BuildTools/VC/Tools/MSVC/14.28.29333/bin/HostX64/x64/cl.exe @bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj.params
2021-02-03T04:44:36.3264090Z Execution platform: @rbe_windows_msvc_cl//config:platform
2021-02-03T04:44:36.3266420Z test/common/http/http2/codec_impl_test.cc(44): error C2220: the following warning is treated as an error
2021-02-03T04:44:36.4959289Z test/common/http/http2/codec_impl_test.cc(44): warning C4455: 'operator ""s': literal suffix identifiers that do not start with an underscore are reserved

File: test/common/http/http2/codec_impl_test.cc
Comment (KBaichoo) [reply test/common/http/http2/codec_impl_test.cc]: I've now replaced this with `EndsWith` instead, that way it has to grok the remaining bytes. I needed to make some adjustments (include the newline, etc. so it makes me confident it is grokking those chars.)

See: https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s about the new C++14 string_literal.

🤞 I can appease the msvc compiler; otherwise, I'll escape as you suggest (It seems like your reasoning for escaping `\0` is tied to this?)

Comment (antoniovicente) [reply test/common/http/http2/codec_impl_test.cc]: The request to escape \0 goes beyond this test.  I would prefer to not have null characters in logs.

File: source/common/common/utility.cc
Comment (KBaichoo) [reply source/common/common/utility.cc:473]: Done.

File: test/common/http/http2/codec_impl_test.cc
Comment (KBaichoo) [reply test/common/http/http2/codec_impl_test.cc]: sg, done.

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review @antoniovicente !

File: source/common/http/http2/codec_impl.cc
Comment (antoniovicente) [inline source/common/http/http2/codec_impl.cc:1431]: nit: I think you're missing spaces in this line

os << spaces << "ConnectionImpl::StreamImpl "

File: test/common/http/http2/codec_impl_test.cc
Comment (antoniovicente) [inline test/common/http/http2/codec_impl_test.cc:980]: nit: passing in a non-zero number of spaces here may make testing easier/more comprehensive.  suggestion: set to 2 instead of 0.

**Review by antoniovicente — DISMISSED** [review]

File: test/common/http/http2/codec_impl_test.cc
Comment (KBaichoo) [reply test/common/http/http2/codec_impl_test.cc:980]: Good idea. It now generates 2 spaces for all of these tests. Note: the `spacesForLevel` generates min(12, 2*i) spaces for i in [0, inf).

File: source/common/http/http2/codec_impl.cc
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc:1431]: Done.

**Comment by KBaichoo** [comment]
@antoniovicente since you're the maintainer on-call can you assign someone from @envoyproxy/senior-maintainers who is a non-googler? Thanks!

**Comment by KBaichoo** [comment]
PTAL @snowp . @antoniovicente suggested you'd be a good senior-maintainer to look at this.

Thanks!

/assign @snowp

**Review by antoniovicente — APPROVED** [review]

**Review by snowp — APPROVED** [review]
Thanks this LGTM!

**Comment by KBaichoo** [comment]
/retest 

Windows seems like a flake

**Comment by moderation** [comment]
I'm getting a compilation error on RHEL 8.3 with Clang 10

image

`no member named 'for_each_n' in namespace 'std'`

Is this a Clang feature issue or something else @KBaichoo ?

**Comment by KBaichoo** [comment]
It is a C++17 feature: https://en.cppreference.com/w/cpp/algorithm/for_each_n

What's the procedure then?

**Comment by moderation** [comment]
This appears to be the first introduction of a C++17 feature and it breaks Clang 10. The current build requirements are outlined at https://github.com/envoyproxy/envoy/blob/main/docs/root/start/building.rst. Will likely need the @envoyproxy/senior-maintainers to chime in

**Comment by jrajahalme** [comment]
`for_each_n` also breaks build on GCC 7. Apparently it is not implemented in libstdc++, maybe libc++ works. Build docs state that both are supported, though. Maybe code the loop without using `for_each_n`?

**Comment by snowp** [comment]
Yeah I think the simplest thing to do here is to just avoid the use of std::for_each_n since it seems to be poorly supported. @lizan do you have any thoughts here? Is this just due to std lib implementations poorly supporting C++17 features?

**Comment by antoniovicente** [comment]
IIRC there are other places where the Envoy codebase depends on C++ 17.  Is there some compiler flag that can be used to disallow for_each_n in c++ 17 if we're not ready to support it yet?

**Comment by KBaichoo** [comment]
We use [structured bindings](https://en.cppreference.com/w/cpp/language/structured_binding) in https://github.com/envoyproxy/envoy/blob/de8c5059baf575aaafdcc42572f7b019824987ba/source/common/upstream/subset_lb.cc#L137 which is c++17

**Comment by snowp** [comment]
Language/library support in compilers tend to be on a per feature basis, so I can totally believe that some of the older standard library versions don't support this feature while still supporting structured bindings. I think if we want to maintain support for GCC-7 we would probably want to avoid using this, similar to how we avoid using std::filesystem due to spotty support on some platforms (e.g. mobile). 

>Is there some compiler flag that can be used to disallow for_each_n in c++ 17 if we're not ready to support it yet?
I doubt it, probably our best bet would be a lint check

**Comment by KBaichoo** [comment]
See: #15093 for a lint of this as you suggested.

**Comment by lizan** [comment]
It must be due to older libstdc++, libstdc++-9 has support for that. @moderation can you try with gcc-toolset-9?

**Comment by moderation** [comment]
@lizan I have gcc-toolset-10 in my RHEL 8 environment and will try that but I recall moving to the clang 10 for some reason. Will try and report back

**Comment by moderation** [comment]
@lizan @KBaichoo With [no config flag](https://github.com/envoyproxy/envoy/blob/main/bazel/README.md#config-flag-choices)  and therefore gcc 10 + libstdc++ and a very limited set of extensions enabled the first failure was in the quic code (I forgot to comment that out).
image

I commented that extension out and the next problem I hit was 
image

So with the addition of the single C++17 feature it doesn't look like there is a way to build on RHEL 8.3 with either GCC 10 or Clang 10

**Comment by lizan** [comment]
You can use libstdc++ >= 9 with clang, no need use gcc, but that gcc 10 issue is a one issue that we should address.

What's your output of `clang -x c++ -v -c /dev/null`? If you see it picks libstdc++-8, you might need symlink `ln -s /opt/rh/gcc-toolset-9/ /usr/lib/gcc/x86_64-redhat-linux/9` to let clang to pick a newer libstc++.

**Comment by rgs1** [comment]
This broke for me too in ubuntu 18 fwiw. Are we planning to revert or document/fix fwd?

**Comment by rgs1** [comment]
Ah nvm #15093 is the fix, i'll wait for that (or add it to our repo in the meanwhile).

**Comment by moderation** [comment]
Looking forward to #15093 too as we are still blocked with Clang 10 on RHEL8.3 (and GCC is broken too but we'd rather stick with Clang)

**Comment by antoniovicente** [comment]
PR #15093 seems stuck in review.  It would be good to unblock folks using ubuntu 18 and RHEL8.3.  What is the recommended path forward?

**Comment by rgs1** [comment]
I'd kindly suggest we revert (PR here https://github.com/envoyproxy/envoy/pull/15153) for now, so we can unblock people who need to track master closely.

**Comment by antoniovicente** [comment]
PR #15093 was merged, ubuntu 18 and RHEL8.3 builds should work again.

**Pull request merged**
